// dingtalk-design-libs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("dingtalk-jsapi/entry/union");
var openTryoutModal_1 = require("./biz/openTryoutModal");
Object.defineProperty(exports, "openTryoutSku", { enumerable: true, get: function () { return openTryoutModal_1.openTryoutSku; } });
var getData_1 = require("./biz/CRM/getData");
Object.defineProperty(exports, "getDataFromCRM", { enumerable: true, get: function () { return getData_1.getDataFromCRM; } });
var contactAdminToUseApp_1 = require("./biz/contactAdminToUseApp");
Object.defineProperty(exports, "contactAdminToUseApp", { enumerable: true, get: function () { return contactAdminToUseApp_1.contactAdminToUseApp; } });
var inviteToUseApp_1 = require("./biz/inviteToUseApp");
Object.defineProperty(exports, "inviteToUseApp", { enumerable: true, get: function () { return inviteToUseApp_1.inviteToUseApp; } });
var openAuthMiniApp_1 = require("./biz/openAuthMiniApp");
Object.defineProperty(exports, "openAuthMiniApp", { enumerable: true, get: function () { return openAuthMiniApp_1.openAuthMiniApp; } });
var openUnifyNav_1 = require("./biz/openUnifyNav");
Object.defineProperty(exports, "openUnifyNav", { enumerable: true, get: function () { return openUnifyNav_1.openUnifyNav; } });
var openSkuForCoolApp_1 = require("./biz/openSkuForCoolApp");
Object.defineProperty(exports, "openSkuForCoolApp", { enumerable: true, get: function () { return openSkuForCoolApp_1.openSkuForCoolApp; } });
var openSkuForCoolAppV2_1 = require("./biz/openSkuForCoolAppV2");
Object.defineProperty(exports, "openSkuForCoolAppV2", { enumerable: true, get: function () { return openSkuForCoolAppV2_1.openSkuForCoolAppV2; } });
var openAuth_1 = require("./biz/openAuth");
Object.defineProperty(exports, "openAuth", { enumerable: true, get: function () { return openAuth_1.openAuth; } });
